File: src/models/CanonicalField.ts
import { Schema, model, Document } from 'mongoose';

export interface ICanonicalField extends Document {
  field_name: string;
  display_name: string;
  field_type: 'String' | 'Number' | 'Date' | 'Boolean' | 'Email' | 'Phone';
  is_active: boolean;
  is_required: boolean;
  version: string;
  createdAt: Date;
  updatedAt: Date;
}

const CanonicalFieldSchema = new Schema<ICanonicalField>(
  {
    field_name: {
      type: String,
      required: [true, 'Field name is required'],
      unique: true,
      trim: true,
      lowercase: true,
      match: [/^[a-z][a-z0-9_]*$/, 'Field name must start with a letter and contain only lowercase letters, numbers, and underscores']
    },
    display_name: {
      type: String,
      required: [true, 'Display name is required'],
      trim: true
    },
    field_type: {
      type: String,
      enum: ['String', 'Number', 'Date', 'Boolean', 'Email', 'Phone'],
      required: [true, 'Field type is required']
    },
    is_active: {
      type: Boolean,
      default: true
    },
    is_required: {
      type: Boolean,
      default: false
    },
    version: {
      type: String,
      default: 'v1'
    }
  },
  {
    timestamps: true,
    collection: 'canonical_fields'
  }
);

// Index for faster queries
CanonicalFieldSchema.index({ field_name: 1 });
CanonicalFieldSchema.index({ is_active: 1 });

export const CanonicalField = model<ICanonicalField>('CanonicalField', CanonicalFieldSchema);

File: backend/bank-leads-java/bank-leads-backend/src/main/java/com/bankleads/bank_leads_backend/model/CanonicalField.java
package com.bankleads.bank_leads_backend.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Document(collection = "canonical_fields")
public class CanonicalField {
    
    @Id
    private String id;
    
    @Indexed(unique = true)
    private String fieldName;
    
    private String displayName;
    
    private FieldType fieldType;
    
    @Indexed
    private Boolean isActive;
    
    private Boolean isRequired;
    
    private String version;
    
    @CreatedDate
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    public enum FieldType {
        String, Number, Date, Boolean, Email, Phone
    }
}
